# Assignment 1: Neural Network and Backpropagation from Scratch

**Course:** Deep Learning for Perception  
**Semester:** Spring 2026  

**Student Name:**  Syeda Maham Mohsin
**Student ID:**    23K-0594

---

## Instructions
- Do NOT use deep learning libraries (TensorFlow, PyTorch, Keras).
- You may use `numpy` and `math`.
- Follow the steps carefully.
- Show intermediate values clearly.
- Do not change the given architecture.
- Compute values of x1 and x2 from your ID. For instance if the ID is Student ID: 23K-1365

Digits → 1, 3, 6, 5

x1=0.13

x2=0.65

In [ ]:
import numpy as np
import math


## Problem Description

We implement a small feedforward neural network with:
- 2 input neurons
- 1 hidden layer with 2 neurons (sigmoid activation)
- 1 output layer with 2 neurons (sigmoid activation)

We will:
1. Perform forward propagation
2. Compute Mean Squared Error
3. Apply backpropagation
4. Update weights using gradient descent


In [ ]:
def sigmoid(z):
    return 1 / (1 + np.exp(-z))

def sigmoid_derivative(z):
    s = sigmoid(z)
    return s * (1 - s)


In [ ]:
# Inputs
x1 = 0.05
x2 = 0.94

# Targets
t1 = 0.01
t2 = 0.99

# Weights (Input → Hidden)
w1, w2 = 0.15, 0.20
w3, w4 = 0.25, 0.30

# Weights (Hidden → Output)
w5, w6 = 0.40, 0.45
w7, w8 = 0.50, 0.55

# Biases
b1 = 0.35
b2 = 0.60

# Learning rate
eta = 0.5


## Forward Propagation

Compute activations of the hidden layer and output layer.


In [ ]:
# Hidden layer
z_h1 = x1*w1 + x2*w2 + b1
z_h2 = x1*w3 + x2*w4 + b1

a_h1 = sigmoid(z_h1)
a_h2 = sigmoid(z_h2)

# Output layer
z_o1 = a_h1*w5 + a_h2*w6 + b2
z_o2 = a_h1*w7 + a_h2*w8 + b2

y1 = sigmoid(z_o1)
y2 = sigmoid(z_o2)

print("Hidden activations:", a_h1, a_h2)
print("Output predictions:", y1, y2)


Hidden activations: 0.633090927809839 0.6557699821438607
Output predictions: 0.7592062758542619 0.781976009039242


## Loss Computation

We use Mean Squared Error (MSE).


In [ ]:
E1 = 0.5 * (t1 - y1)**2
E2 = 0.5 * (t2 - y2)**2
E_total = E1 + E2

print("Total Error:", E_total)


Total Error: 0.30229201229732694


## Backpropagation

Compute gradients using the chain rule.


In [ ]:
# Output layer gradients
dE_dy1 = y1 - t1
dE_dy2 = y2 - t2

dy1_dz1 = sigmoid_derivative(z_o1)
dy2_dz2 = sigmoid_derivative(z_o2)

# Gradients for weights
dE_dw5 = dE_dy1 * dy1_dz1 * a_h1
dE_dw6 = dE_dy1 * dy1_dz1 * a_h2
dE_dw7 = dE_dy2 * dy2_dz2 * a_h1
dE_dw8 = dE_dy2 * dy2_dz2 * a_h2

print("Gradients:", dE_dw5, dE_dw6, dE_dw7, dE_dw8)


Gradients: 0.08671065161429527 0.08981686510262013 -0.022453147460061978 -0.023257480817006716


## Weight Update

Update weights using gradient descent.


In [ ]:
w5 -= eta * dE_dw5
w6 -= eta * dE_dw6
w7 -= eta * dE_dw7
w8 -= eta * dE_dw8

print("Updated weights:")
print(w5, w6, w7, w8)


Updated weights:
0.3566446741928524 0.40509156744868996 0.511226573730031 0.5616287404085034


## Some more tasks

1. Did the total error decrease after one iteration?
2. What would happen if the learning rate were too large?
3. Repeat this with a larger learning rate and one more input. You may select x3 with a random value between .1 and .5


In [ ]:
#Question 1: Forward pass again after weight update

#Hidden layer
z_h1_new = x1*w1 + x2*w2 + b1
z_h2_new = x1*w3 + x2*w4 + b1

a_h1_new = sigmoid(z_h1_new)
a_h2_new = sigmoid(z_h2_new)

#Output layer with UPDATED weights
z_o1_new = a_h1_new*w5 + a_h2_new*w6 + b2
z_o2_new = a_h1_new*w7 + a_h2_new*w8 + b2

y1_new = sigmoid(z_o1_new)
y2_new = sigmoid(z_o2_new)

#New error
E1_new = 0.5 * (t1 - y1_new)**2
E2_new = 0.5 * (t2 - y2_new)**2
E_total_new = E1_new + E2_new

print("Old Error:", E_total)
print("New Error:", E_total_new)


Old Error: 0.30229201229732694
New Error: 0.29392309246179166


ANS 1) Yes.

Total error before update = 0.302

Total error after update = 0.293

Since the error reduced after updating the weights, this confirms that gradient descent is reducing error as expected, while our system is learning.

ANS 2) If the learning rate is too large:
 The Weight updates become excessively big and the optimization may overshoot the minimum and the loss may:

1) increase instead of decrease

2) oscillate

3) diverge completely

4) Training becomes unstable and may fail to converge


In [ ]:
#ANS 3

# Old Inputs
x1 = 0.05
x2 = 0.94

# New input
x3 = 0.20

# Targets
t1 = 0.01
t2 = 0.99

# Weights (Input → Hidden)
w1, w2, w3 = 0.15, 0.20, 0.25
w4, w5, w6 = 0.30, 0.40, 0.45

# Weights (Hidden → Output)
w7, w8 = 0.50, 0.55
#New weights for x3
w9, w10 = 0.60, 0.65

# Biases
b1 = 0.35
b2 = 0.60


eta_large = 2.5 #larger learning rate


# Hidden layer
z_h1 = x1*w1 + x2*w2 + x3*w3 + b1
z_h2 = x1*w4 + x2*w5 + x3*w6 + b1

a_h1 = sigmoid(z_h1)
a_h2 = sigmoid(z_h2)

# Output layer
z_o1 = a_h1*w7 + a_h2*w8 + b2
z_o2 = a_h1*w9 + a_h2*w10 + b2

y1 = sigmoid(z_o1)
y2 = sigmoid(z_o2)

print("Hidden activations:", a_h1, a_h2)
print("Output predictions:", y1, y2)

E1 = 0.5 * (t1 - y1)**2
E2 = 0.5 * (t2 - y2)**2
E_total = E1 + E2

print("Error with larger learning rate and extra input:", E_total)

# Output layer gradients
dE_dy1 = y1 - t1
dE_dy2 = y2 - t2

dy1_dz1 = sigmoid_derivative(z_o1)
dy2_dz2 = sigmoid_derivative(z_o2)

# Gradients for weights
dE_dw7 = dE_dy1 * dy1_dz1 * a_h1
dE_dw8 = dE_dy1 * dy1_dz1 * a_h2
dE_dw9 = dE_dy2 * dy2_dz2 * a_h1
dE_dw10 = dE_dy2 * dy2_dz2 * a_h2

print("Gradients:", dE_dw7, dE_dw8, dE_dw9, dE_dw10)

Hidden activations: 0.6446261036324369 0.6965663330376879
Output predictions: 0.7867465078237356 0.8083839930077766
Error with larger learning rate and extra input: 0.3181597557060839
Gradients: 0.08400752173228228 0.09077636017359574 -0.018134747047614 -0.01959593969332223


In [ ]:
print("Old weights:")
print(w7, w8, w9, w10)

w7 -= eta * dE_dw7
w8 -= eta * dE_dw8
w9 -= eta * dE_dw9
w10 -= eta * dE_dw10

print()

print("Updated weights:")
print(w7, w8, w9, w10)

Old weights:
0.5 0.55 0.6 0.65

Updated weights:
0.45799623913385884 0.5046118199132021 0.609067373523807 0.6597979698466612


3) With a larger learning rate and extra input, weight updates become larger and error changes more aggressively. Learning may become faster but less stable.